src/Raven.Server/Documents/Replication/OutgoingDocumentReplication.cs

```diff
@@ -1,90 +1,186 @@
+                    // TODO [ppekrol] GC?
+                }
+                catch (OperationCanceledException)
+                {
+                    return;
+                }
+                catch (Exception e)
+                {
+                    _log.WarnException($"Exception occured for '{_replicationUniqueName}'.", e);
+                }
+
+                if (HasMoreDocumentsToSend)
+                    continue;
+
+                try
+                {
+                    //if this returns false, this means canceled token is activated                    
+                    if (await _waitForChanges.WaitAsync() == false)
+                        //thus, if code reaches here, cancellation token source has "cancel" requested
+                        return;
+                }
+                catch (OperationCanceledException)
+                {
+                    return;
+                }
+            }
+        }
+
+        public bool HasMoreDocumentsToSend
+        {
+            get
+            {
+                using(_context.OpenReadTransaction())
+                    return DocumentsStorage.ReadLastEtag(
+                        _context.Transaction.InnerTransaction) < _lastSentEtag;
+            }
+        }
+
+        private async Task ExecuteReplicationOnce()
+        {
+            Debug.Assert(_isInitialized);
 
             //just for shorter code
             var documentStorage = _database.DocumentsStorage;
-            DocumentsOperationContext context;
-            using (documentStorage.ContextPool.AllocateOperationContext(out context))
-            using (context.OpenReadTransaction())
+            using (_context.OpenReadTransaction())
             {
                 //TODO: make replication batch size configurable
                 //also, perhaps there should be timers/heuristics
                 //that would dynamically resize batch size
                 var replicationBatch =
                     documentStorage
-                        .GetDocumentsAfter(context, lastSendEtag, 0, 1024)
-                        .Where(x => !x.Key.ToString().StartsWith(SystemDocumentPrefix))
-                        .ToArray();
-
-                if (replicationBatch.Length == 0)
-                {
-                    _hasMoreDocumentsToSend = true;
+                        .GetDocumentsAfter(_context, _lastSentEtag, 0, 1024)
+                        .Where(x => !x.Key.ToString().StartsWith("Raven/"))
+                        .ToList();
+                        //the filtering here will need to be reworked -> it is not efficient
+                        //TODO: do not forget to make version of GetDocumentsAfter with a prefix filter	
+                        //alternatively 1 -> create efficient StartsWith() for LazyString				
+                        //alternatively 2 -> create a "filter system" that would abstract the logic -> what documents 
```

Comment:
We'll need to filter elsewhere, not here.
Note that we also need to make sure that the size of the batch is adjusting based on how much time it takes to send it.